Modify rule S6427: LaYC format (#2449)
SonarSource/rspec: rules/S6427/cfamily/rule.adoc

== Why is this an issue?

The class `std::optional<T>` either stores a value of type `T` or is empty.

> One way to access the value of a non-empty optional is the `operator*`, making an optional look like a pointer. However, the similarity ends there: an optional object models an object, not a pointer. If the optional is uninitialized, assigning a value to the dereferenced optional is undefined behavior.

Reviewer: Technically, just calling `operator*` on an empty optional is UB. This suggestion removes the notion of "assignment" since it's already past the point of UB. I tried to keep it short but maybe there's a better alternative.

```suggestion
One way to access the value of a non-empty optional is the `operator*`, making an optional look like a pointer. However, the similarity ends there: an optional object models an object, not a pointer. Furthermore, dereferencing an empty optional is undefined behavior.
```

Author: I have rewritten this paragraph a bit in https://github.com/SonarSource/rspec/pull/2449/commits/ebf7f7963022f73f824203a3f1359f9f29c452f7
WDYT?

Another way to access the value of a non-empty optional is the function `value()`. But assigning a value to the optional object through this function will throw an exception (`std::bad_optional_access`) if the optional has no value, and the assignment will not happen.

For the assignment of an optional to happen correctly, whatever its state, it is better to:

> * assign the value directly with the `operator=`: e.g. `std::optional<int> opt = 3;`

Reviewer: Nit: the example may give the impression we need a new variable.

```suggestion
* assign the value directly with the `operator=`: e.g. `myOptionalInteger = 3;`
```

== Why is this an issue?

> The class `std::optional` stores an optional value: a `std::optional<T>` can either contain a value of type `T` or be empty.

Reviewer: I feel this sentence is very repetitive. Maybe something like this:
```suggestion
The class `std::optional<T>` is either in an empty state or it stores an value of type `T`.
```

Author: Modified, let me know what you think

Reviewer: Much better! 👍

> One way to access the value of a non-empty optional is the `operator*`, making an optional look like a pointer. However, the similarity ends there: an optional object models an object, not a pointer. Assigning a value to the dereferenced optional or the result of the function `value()` is undefined behavior if the optional has no value. That is why the preferred way to assign a value to an optional is to assign it directly: it will work in any case.

Reviewer: > Assigning a value to [...] the result of the function value() is undefined behavior if the optional has no value.

Technically, this is well defined: `.value()` will throw and the assignement will never happen.

The connection with the pointer semantics works relatively well, I think, when using `*` or `->` to access the value: it is undefined on `nullptr` and it is undefined on `nullopt`.

The connection with pointers stops when `.value()` is involved: there is simply no counterpart for pointers.

> the preferred way [...] is to assign it directly

This could be more explicit and refer to the `operator=` of the optional, or its `emplace` function -- the later is also interesting when the move or copy operation is expensive or forbidden.

Author: Updated in https://github.com/SonarSource/rspec/pull/2449/commits/3533d1f1e18dbf179e7b464c7cc93208cdd7321d

== How to fix it

=== Code examples

==== Noncompliant code example

[source,cpp,diff-id=1,diff-type=noncompliant]
----
void g(std::optional<int> &val, bool b) {
    if (b) {

>       *val = 314; // Noncompliant, it is undefined behavior if the optional is empty

Reviewer: ```suggestion
      *val = 314; // Noncompliant; the behavior is undefined if the optional is empty.
```

    } else {

>       val.value() = 42; // Noncompliant, it will throw if the optional is empty

Reviewer: ```suggestion
      val.value() = 42; // Noncompliant; it will throw if the optional is empty.
```

== Why is this an issue?

The class `std::optional<T>` either stores a value of type `T` or is empty.

One way to access the value of a non-empty optional is the `operator*`, making an optional look like a pointer. However, the similarity ends there: an optional object models an object, not a pointer. If the optional is uninitialized, assigning a value to the dereferenced optional is undefined behavior.

Another way to access the value of a non-empty optional is the function `value()`. But assigning a value to the optional object through this function will throw an exception (`std::bad_optional_access`) if the optional has no value, and the assignment will not happen.

For the assignment of an optional to happen correctly, whatever its state, it is better to:

* assign the value directly with the `operator=`: e.g. `std::optional<int> opt = 3;`
* use the `emplace` function (for example, when the move or copy operation is expensive or forbidden).

== How to fix it

> The rule raises an issue when the `operator=` or the `value()` function initializes an optional.

Reviewer: ```suggestion
The rule raises an issue when the `operator*` or the `value()` function are used to assign a new value to an optional.
```
